src/Auth/Source/SQL2.php

```diff
@@ -0,0 +1,622 @@
+
+
+    /**
+     * Authenticate using the optional password_verify() support against a hash retrieved from the database.
+     *
+     * @param string $queryname   Name of the auth query being processed
+     * @param array $queryConfig  Configuration from authsources.php for this auth query
+     * @param array $data         Result data from the database query
+     * @param string $password    Password to verify with password_verify()
+     * @return bool  True if password_verify() password verification succeeded, false otherwise
+     */
+    protected function authenticatePasswordVerifyHash(
+        string $queryname,
+        array $queryConfig,
+        array $data,
+        string $password,
+    ): bool {
+        // If password_verify_hash_column is not set, we are not using password_verify()
+        if (!array_key_exists('password_verify_hash_column', $queryConfig)) {
+            Logger::error(sprintf(
+                'sqlauth:%s: authenticatePasswordVerifyHash() called but configuration for ' .
+                '"password_verify_hash_column" not found in query config for query %s.',
+                $this->authId,
+                $queryname,
+            ));
+            throw new Error\Error('WRONGUSERPASS');
+        } elseif (count($data) < 1) {
+            // No rows returned, password_verify() cannot succeed
+            return false;
+        }
+
+        /* This is where we need to run password_verify() if we are using password_verify() to
+            * authenticate hashed passwords that are only stored in the database. */
+        $hashColumn = $queryConfig['password_verify_hash_column'];
+        if (!array_key_exists($hashColumn, $data[0])) {
+            Logger::error('sqlauth:' . $this->authId . ': Auth query ' . $queryname .
+                            ' did not return expected hash column \'' . $hashColumn . '\'');
+            throw new Error\Error('WRONGUSERPASS');
+        }
+
+        $validPasswordHashFound = false;
+        $passwordHash = null;
+        foreach ($data as $row) {
+            if ((!array_key_exists($hashColumn, $row)) || is_null($row[$hashColumn])) {
+                Logger::error(sprintf(
+                    'sqlauth:%s: column `%s` must be in every result tuple.',
+                    $this->authId,
+                    $hashColumn,
+                ));
+                throw new Error\Error('WRONGUSERPASS');
+            }
+            if (($passwordHash === null) && (strlen($row[$hashColumn]) > 0)) {
+                $passwordHash = $row[$hashColumn];
+                $validPasswordHashFound = true;
+            } elseif ($passwordHash != $row[$hashColumn]) {
+                Logger::error(sprintf(
+                    'sqlauth:%s: column %s must be THE SAME in every result tuple.',
+                    $this->authId,
+                    $hashColumn,
+                ));
+                throw new Error\Error('WRONGUSERPASS');
+            } elseif (strlen($row[$hashColumn]) === 0) {
```

Comment:
I think this strlen test is better as just a raw `if` near the top of this foreach (above line 448 for example). The `strlen($row[$hashColumn])` must always be `> 0` regardless of anything else that is happening. 

I can analyse that if `$passwordHash === null` and `strlen == 0` and `$passwordHash != $row[$hashColumn]` will not be true when `$passwordHash == null` and strlen == 0, so the final elseif will run. But it is simpler logic to just lift the `strlen($row[$hashColumn]) > 0` test out of this if/elseif/elsif block. If it is a single `if()` for every result inspected in the result set it is very clear it always runs and negates bad data without much thought needed.